### PR TITLE
Add the protocols RealFunctions and Real

### PR DESCRIPTION
**Pull request by stephentyrone**
RealFunctions refines ElementaryFunctions, adding operations that only make sense for types that model the real numbers (as opposed to complex, quaternions, matrices, etc, which all can plausibly conform to ElementaryFunctions). SIMD types whose elements conform also conform, beecause the operations are interpreted componentwise.

We also introduce Real, which restricts to RealFunctions + FloatingPoint, effectively removing the SIMD types. This does not add any new conformances; rather it is intended as a practical protocol for use in writing generic code (which would otherwise be constrained to RealFunctions & FloatingPoint more often than not).

This is a small modification of what we pitched and approved as SE-0246, so I'm going to re-pitch and run a minor revision proposal through SE for it.

### CODE REVIEW

**Comment by stephentyrone** [comment]
@swift-ci smoke test

**Comment by stephentyrone** [comment]
@swift-ci test source compatibility

File: stdlib/public/Darwin/CoreGraphics/CGFloat.swift.gyb [hunk 1]
```diff
@@ -510,12 +510,12 @@ public func %=(lhs: inout CGFloat, rhs: CGFloat) {
 }
 
 //===----------------------------------------------------------------------===//
-// Real conformance
+// RealProtocol conformance
```
Comment (rxwei) [inline stdlib/public/Darwin/CoreGraphics/CGFloat.swift.gyb]: When I looked at this diff I thought the `Real` protocol got renamed to `RealProtocol`.

Comment (stephentyrone) [reply stdlib/public/Darwin/CoreGraphics/CGFloat.swift.gyb]: Fixed.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**